internal/cimfs/mount_cim.go

```diff
@@ -0,0 +1,84 @@
+package cimfs
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/Microsoft/go-winio/pkg/guid"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"github.com/pkg/errors"
+)
+
+type MountError struct {
+	Cim        string
+	Op         string
+	VolumeGUID guid.GUID
+	Err        error
+}
+
+func (e *MountError) Error() string {
+	s := "cim " + e.Op
+	if e.Cim != "" {
+		s += " " + e.Cim
+	}
+	s += " " + e.VolumeGUID.String() + ": " + e.Err.Error()
+	return s
+}
+
+type cimInfo struct {
+	// path to the cim
+	path string
+	// Unique GUID assigned to a cim.
+	cimID guid.GUID
+	// ref count for number of times this cim was mounted.
+	refCount uint32
+}
+
+// map for information about cims mounted on the host
+var hostCimMounts = make(map[string]*cimInfo)
+
+// Mount mounts the cim at path `cimPath` and returns the mount location of that cim.
+// If this cim is already mounted then nothing is done.
+func Mount(cimPath string) (string, error) {
+	if _, ok := hostCimMounts[cimPath]; !ok {
+		layerGUID, err := guid.NewV4()
+		if err != nil {
+			return "", &MountError{Cim: cimPath, Op: "Mount", Err: err}
+		}
+		if err := winapi.CimMountImage(filepath.Dir(cimPath), filepath.Base(cimPath), hcsschema.CimMountFlagCacheFiles, &layerGUID); err != nil {
+			return "", &MountError{Cim: cimPath, Op: "Mount", VolumeGUID: layerGUID, Err: err}
+		}
+		hostCimMounts[cimPath] = &cimInfo{cimPath, layerGUID, 0}
```

Comment:
Should we use a lock for any operations on this?

Reply:
Good point. Added a mutex.